### PR TITLE
Re-factored to use Elasticsearch Bulk API

### PR DESCRIPTION
**Pull request by efontana**
This is a re-factoring of the Elasticsearch outputter to use the Bulk API, this will close this issue:

https://github.com/Cimpress-MCP/TimberWinR/issues/29

@chrisbaldauf @ryanbreen please review.

### CODE REVIEW

File: TimberWinR/Parser.cs [hunk 1]
```diff
@@ -512,9 +512,19 @@ public class ElasticsearchOutputParameters
...
+        [JsonProperty(PropertyName = "idle_flush_time")]
+        public int IdleFlushTimeInSeconds { get; set; }
+        [JsonProperty(PropertyName = "max_queue_size")]
+        public int MaxQueueSize { get; set; }
+        [JsonProperty(PropertyName = "queue_overflow_discard_oldest")]
+        public bool QueueOverflowDiscardOldest { get; set; }
 
         public ElasticsearchOutputParameters()
         {
+            FlushSize = 5000;
```
Comment (ryanbreen) [inline TimberWinR/Parser.cs]: The docs say the default flushsize is 50000.  Which one is right?

File: TimberWinR/mdocs/ElasticsearchOutput.md [hunk 2]
```diff
@@ -7,11 +7,15 @@ The following parameters are allowed when configuring the Redis output.
 
 | Parameter     |   Type   |  Description                                                | Details               |  Default |
 | :-------------|:---------|:------------------------------------------------------------| :---------------------------  | :-- |
-| *threads*     | string   | Location of log files(s) to monitor                         | Number of worker theads to send messages | 1 |
-| *interval*    | integer  | Interval in milliseconds to sleep during batch sends        | Interval       | 5000 |
-| *index*       | string   | The index name to use                                       | index used/created | logstash-yyyy.dd.mm |
-| *host*        | [string] | The hostname(s) of your Elasticsearch server(s) | IP or DNS name |  |
-| *port*        | integer  | Redis port number                                           | This port must be open  | 9200  |
+| *flush_size*                    | integer  | Maximum number of messages before flushing           |   | 50000  |
```
Comment (ryanbreen) [inline TimberWinR/mdocs/ElasticsearchOutput.md]: Do users need to take any action as a result of these new config params, or do we expect these defaults to be reasonable?

File: TimberWinR/Parser.cs [hunk 3]
```diff
@@ -512,9 +512,19 @@ public class ElasticsearchOutputParameters
...
+            IdleFlushTimeInSeconds = 1;
```
Comment (ryanbreen) [inline TimberWinR/Parser.cs]: Docs say IdleFlushTimeInSeconds default is 10.

File: TimberWinR/Outputs/Elasticsearch.cs [hunk 4]
```diff
@@ -2,52 +2,91 @@
...
 using RapidRegex.Core;
 using RestSharp;
+using System.Text.RegularExpressions;
+using Elasticsearch.Net.Serialization;
+using Newtonsoft.Json;
 
 namespace TimberWinR.Outputs
 {
-    using System.Text.RegularExpressions;
+    public class Person
```
Comment (chrisbaldauf) [inline TimberWinR/Outputs/Elasticsearch.cs]: Is this really supposed to be here?

File: TimberWinR/Outputs/Elasticsearch.cs [hunk 5]
```diff
@@ -2,52 +2,91 @@
...
+    {
+        public string Firstname { get; set; }
+        public string Lastname { get; set; }
+    }
 
     public partial class ElasticsearchOutput : OutputSender
     {
         private TimberWinR.Manager _manager;
         private readonly int _port;
```
Comment (chrisbaldauf) [inline TimberWinR/Outputs/Elasticsearch.cs]: Do all of the hosts need to be on the same port?

File: TimberWinR/Outputs/Elasticsearch.cs [hunk 6]
```diff
@@ -2,52 +2,91 @@
...
+            _protocol = parameters.Protocol;
+            _timeout = parameters.Timeout;
             _manager = manager;
-            _port = eo.Port;
-            _interval = eo.Interval;
-            _host = eo.Host;
-            _hostIndex = 0;
+            _port = parameters.Port;
+            _interval = parameters.Interval;
+            _hosts = parameters.Host;
```
Comment (chrisbaldauf) [inline TimberWinR/Outputs/Elasticsearch.cs]: is parameters.Host a list? should it be .Hosts?

File: TimberWinR/Outputs/Elasticsearch.cs [hunk 7]
```diff
@@ -58,16 +97,20 @@ public override JObject ToJson()
             JObject json = new JObject(
                 new JProperty("elasticsearch",
                     new JObject(
-                        new JProperty("host", string.Join(",", _host)),
+                        new JProperty("host", string.Join(",", _hosts)),
```
Comment (chrisbaldauf) [inline TimberWinR/Outputs/Elasticsearch.cs]: Another instance of host vs hosts

File: TimberWinR/Outputs/Elasticsearch.cs [hunk 8]
```diff
@@ -58,16 +97,20 @@ public override JObject ToJson()
...
                         new JProperty("errors", _errorCount),
                         new JProperty("sentMmessageCount", _sentMessages),
```
Comment (chrisbaldauf) [inline TimberWinR/Outputs/Elasticsearch.cs]: Typo?

File: TimberWinR/Outputs/Elasticsearch.cs [hunk 9]
```diff
@@ -77,137 +120,145 @@ public override JObject ToJson()
...
-                                        }
-                                        else
-                                        {
-                                            LogManager.GetCurrentClassLogger()
-                                                .Fatal("Unable to connect with any Elasticsearch hosts, {0}",
-                                                    String.Join(",", _host));
-                                            Interlocked.Increment(ref _errorCount);
-                                        }
+                                        // Grab all messages with same index and type 
+                                        var bulkTypeName = this._parameters.GetTypeName(messages[0]);
```
Comment (chrisbaldauf) [inline TimberWinR/Outputs/Elasticsearch.cs]: Does this assume that all of the messages are being sent to the same index and are of the same type? Is there anything enforcing this upstream?

File: TimberWinR/mdocs/ElasticsearchOutput.md [hunk 10]
```diff
@@ -7,11 +7,15 @@ The following parameters are allowed when configuring the Redis output.
...
+| *host*                          | [string] | The hostname(s) of your Elasticsearch server(s) | IP or DNS name |  |
```
Comment (chrisbaldauf) [inline TimberWinR/mdocs/ElasticsearchOutput.md]: What is the delimiter for hosts?

**Comment by chrisbaldauf** [comment]
Did you intend to commit the contents of the packages directory? I'd expect that to be gitignore'd, since it can be resolved with NuGet

**Pull request merged**